<a href="https://colab.research.google.com/github/codingniket/Python-Training/blob/main/17_12_25/17_12_25.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [24]:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("SQL").getOrCreate()
from pyspark.sql.functions import col,substring

In [5]:
sales_data = [
    ("S001","North","Laptop","2024-01-05",75000),
    ("S002","South","Mobile","2024-01-06",30000),
    ("S003","East","Laptop","2024-01-07",72000),
    ("S004","West","Tablet","2024-01-08",25000),
    ("S005","North","Mobile","2024-01-10",28000),
    ("S006","South","Laptop","2024-01-11",76000),
    ("S007","East","Tablet","2024-01-12",24000),
    ("S008","West","Mobile","2024-01-13",32000),
    ("S009","North","Tablet","2024-01-14",26000),
    ("S010","South","Mobile","2024-01-15",31000),
    ("S011","East","Laptop","2024-01-16",74000),
    ("S012","West","Laptop","2024-01-17",78000)
]

sales_cols = ["sales_id","region","product","sale_date","amount"]

df_sales = spark.createDataFrame(sales_data,sales_cols)
df_sales.show()
df_sales.printSchema()

+--------+------+-------+----------+------+
|sales_id|region|product| sale_date|amount|
+--------+------+-------+----------+------+
|    S001| North| Laptop|2024-01-05| 75000|
|    S002| South| Mobile|2024-01-06| 30000|
|    S003|  East| Laptop|2024-01-07| 72000|
|    S004|  West| Tablet|2024-01-08| 25000|
|    S005| North| Mobile|2024-01-10| 28000|
|    S006| South| Laptop|2024-01-11| 76000|
|    S007|  East| Tablet|2024-01-12| 24000|
|    S008|  West| Mobile|2024-01-13| 32000|
|    S009| North| Tablet|2024-01-14| 26000|
|    S010| South| Mobile|2024-01-15| 31000|
|    S011|  East| Laptop|2024-01-16| 74000|
|    S012|  West| Laptop|2024-01-17| 78000|
+--------+------+-------+----------+------+

root
 |-- sales_id: string (nullable = true)
 |-- region: string (nullable = true)
 |-- product: string (nullable = true)
 |-- sale_date: string (nullable = true)
 |-- amount: long (nullable = true)



In [8]:
df_sales.select("sales_id","amount","product").show()

+--------+------+-------+
|sales_id|amount|product|
+--------+------+-------+
|    S001| 75000| Laptop|
|    S002| 30000| Mobile|
|    S003| 72000| Laptop|
|    S004| 25000| Tablet|
|    S005| 28000| Mobile|
|    S006| 76000| Laptop|
|    S007| 24000| Tablet|
|    S008| 32000| Mobile|
|    S009| 26000| Tablet|
|    S010| 31000| Mobile|
|    S011| 74000| Laptop|
|    S012| 78000| Laptop|
+--------+------+-------+



In [9]:
df_sales.select(col("sales_id"),col("amount").alias("sales")).show()

+--------+-----+
|sales_id|sales|
+--------+-----+
|    S001|75000|
|    S002|30000|
|    S003|72000|
|    S004|25000|
|    S005|28000|
|    S006|76000|
|    S007|24000|
|    S008|32000|
|    S009|26000|
|    S010|31000|
|    S011|74000|
|    S012|78000|
+--------+-----+



**SET 1**

In [10]:
sales_data = [
("T001","North","Delhi","Store-01","Laptop","2024-01-01",75000),
("T002","North","Delhi","Store-01","Mobile","2024-01-02",32000),
("T003","North","Chandigarh","Store-02","Tablet","2024-01-03",26000),
("T004","South","Bangalore","Store-03","Laptop","2024-01-01",78000),
("T005","South","Chennai","Store-04","Mobile","2024-01-02",30000),
("T006","South","Bangalore","Store-03","Tablet","2024-01-03",24000),
("T007","East","Kolkata","Store-05","Laptop","2024-01-01",72000),
("T008","East","Kolkata","Store-05","Mobile","2024-01-02",28000),
("T009","East","Patna","Store-06","Tablet","2024-01-03",23000),
("T010","West","Mumbai","Store-07","Laptop","2024-01-01",80000),
("T011","West","Mumbai","Store-07","Mobile","2024-01-02",35000),
("T012","West","Pune","Store-08","Tablet","2024-01-03",27000),
("T013","North","Delhi","Store-01","Laptop","2024-01-04",76000),
("T014","South","Chennai","Store-04","Laptop","2024-01-04",79000),
("T015","East","Patna","Store-06","Mobile","2024-01-04",29000),
("T016","West","Pune","Store-08","Laptop","2024-01-04",77000),
("T017","North","Chandigarh","Store-02","Mobile","2024-01-05",31000),
("T018","South","Bangalore","Store-03","Mobile","2024-01-05",34000),
("T019","East","Kolkata","Store-05","Tablet","2024-01-05",25000),
("T020","West","Mumbai","Store-07","Tablet","2024-01-05",29000),
("T021","North","Delhi","Store-01","Tablet","2024-01-06",28000),
("T022","South","Chennai","Store-04","Tablet","2024-01-06",26000),
("T023","East","Patna","Store-06","Laptop","2024-01-06",74000),
("T024","West","Pune","Store-08","Mobile","2024-01-06",33000)
]
columns = [
"txn_id","region","city","store_id","product","sale_date","amount"
]
df_sales = spark.createDataFrame(sales_data, columns)
df_sales.show(5)
df_sales.printSchema()

+------+------+----------+--------+-------+----------+------+
|txn_id|region|      city|store_id|product| sale_date|amount|
+------+------+----------+--------+-------+----------+------+
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|
|  T004| South| Bangalore|Store-03| Laptop|2024-01-01| 78000|
|  T005| South|   Chennai|Store-04| Mobile|2024-01-02| 30000|
+------+------+----------+--------+-------+----------+------+
only showing top 5 rows
root
 |-- txn_id: string (nullable = true)
 |-- region: string (nullable = true)
 |-- city: string (nullable = true)
 |-- store_id: string (nullable = true)
 |-- product: string (nullable = true)
 |-- sale_date: string (nullable = true)
 |-- amount: long (nullable = true)



1. Select only txn_id , region , product , and amount

In [13]:
mini_data = df_sales.select("txn_id","region","product","amount")
mini_data.show()

+------+------+-------+------+
|txn_id|region|product|amount|
+------+------+-------+------+
|  T001| North| Laptop| 75000|
|  T002| North| Mobile| 32000|
|  T003| North| Tablet| 26000|
|  T004| South| Laptop| 78000|
|  T005| South| Mobile| 30000|
|  T006| South| Tablet| 24000|
|  T007|  East| Laptop| 72000|
|  T008|  East| Mobile| 28000|
|  T009|  East| Tablet| 23000|
|  T010|  West| Laptop| 80000|
|  T011|  West| Mobile| 35000|
|  T012|  West| Tablet| 27000|
|  T013| North| Laptop| 76000|
|  T014| South| Laptop| 79000|
|  T015|  East| Mobile| 29000|
|  T016|  West| Laptop| 77000|
|  T017| North| Mobile| 31000|
|  T018| South| Mobile| 34000|
|  T019|  East| Tablet| 25000|
|  T020|  West| Tablet| 29000|
+------+------+-------+------+
only showing top 20 rows


2. Rename amount to revenue

In [14]:
mini_data = df_sales.select("txn_id","region","product",col("amount").alias("revenue"))
mini_data.show()

+------+------+-------+-------+
|txn_id|region|product|revenue|
+------+------+-------+-------+
|  T001| North| Laptop|  75000|
|  T002| North| Mobile|  32000|
|  T003| North| Tablet|  26000|
|  T004| South| Laptop|  78000|
|  T005| South| Mobile|  30000|
|  T006| South| Tablet|  24000|
|  T007|  East| Laptop|  72000|
|  T008|  East| Mobile|  28000|
|  T009|  East| Tablet|  23000|
|  T010|  West| Laptop|  80000|
|  T011|  West| Mobile|  35000|
|  T012|  West| Tablet|  27000|
|  T013| North| Laptop|  76000|
|  T014| South| Laptop|  79000|
|  T015|  East| Mobile|  29000|
|  T016|  West| Laptop|  77000|
|  T017| North| Mobile|  31000|
|  T018| South| Mobile|  34000|
|  T019|  East| Tablet|  25000|
|  T020|  West| Tablet|  29000|
+------+------+-------+-------+
only showing top 20 rows


3. Create a derived column amount_in_thousands

In [18]:
from pyspark.sql.functions import col
df = df_sales.withColumn('amount_in_thousands', (col('amount') / 1000).cast('int'))

mini_data = df.select("txn_id","region","product",col("amount_in_thousands").alias("revenue (in thousand)"))
mini_data.show()

+------+------+-------+---------------------+
|txn_id|region|product|revenue (in thousand)|
+------+------+-------+---------------------+
|  T001| North| Laptop|                   75|
|  T002| North| Mobile|                   32|
|  T003| North| Tablet|                   26|
|  T004| South| Laptop|                   78|
|  T005| South| Mobile|                   30|
|  T006| South| Tablet|                   24|
|  T007|  East| Laptop|                   72|
|  T008|  East| Mobile|                   28|
|  T009|  East| Tablet|                   23|
|  T010|  West| Laptop|                   80|
|  T011|  West| Mobile|                   35|
|  T012|  West| Tablet|                   27|
|  T013| North| Laptop|                   76|
|  T014| South| Laptop|                   79|
|  T015|  East| Mobile|                   29|
|  T016|  West| Laptop|                   77|
|  T017| North| Mobile|                   31|
|  T018| South| Mobile|                   34|
|  T019|  East| Tablet|           

4. Select distinct combinations of region and product

In [20]:
mini_data.groupBy("region","product").count().show()

+------+-------+-----+
|region|product|count|
+------+-------+-----+
| North| Laptop|    2|
| North| Tablet|    2|
|  East| Tablet|    2|
|  East| Laptop|    2|
| South| Tablet|    2|
| North| Mobile|    2|
|  West| Tablet|    2|
|  East| Mobile|    2|
| South| Mobile|    2|
| South| Laptop|    2|
|  West| Mobile|    2|
|  West| Laptop|    2|
+------+-------+-----+



5. Select all columns but exclude store_id

In [23]:
dropping = mini_data.drop(col("store_id"))
dropping.show()

+------+------+-------+---------------------+
|txn_id|region|product|revenue (in thousand)|
+------+------+-------+---------------------+
|  T001| North| Laptop|                   75|
|  T002| North| Mobile|                   32|
|  T003| North| Tablet|                   26|
|  T004| South| Laptop|                   78|
|  T005| South| Mobile|                   30|
|  T006| South| Tablet|                   24|
|  T007|  East| Laptop|                   72|
|  T008|  East| Mobile|                   28|
|  T009|  East| Tablet|                   23|
|  T010|  West| Laptop|                   80|
|  T011|  West| Mobile|                   35|
|  T012|  West| Tablet|                   27|
|  T013| North| Laptop|                   76|
|  T014| South| Laptop|                   79|
|  T015|  East| Mobile|                   29|
|  T016|  West| Laptop|                   77|
|  T017| North| Mobile|                   31|
|  T018| South| Mobile|                   34|
|  T019|  East| Tablet|           

6. Create a new column sale_year extracted from sale_date


In [28]:
df = df_sales.withColumn('year', substring(col("sale_date"),0,4))
df.show()

+------+------+----------+--------+-------+----------+------+----+
|txn_id|region|      city|store_id|product| sale_date|amount|year|
+------+------+----------+--------+-------+----------+------+----+
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|2024|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|2024|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|2024|
|  T004| South| Bangalore|Store-03| Laptop|2024-01-01| 78000|2024|
|  T005| South|   Chennai|Store-04| Mobile|2024-01-02| 30000|2024|
|  T006| South| Bangalore|Store-03| Tablet|2024-01-03| 24000|2024|
|  T007|  East|   Kolkata|Store-05| Laptop|2024-01-01| 72000|2024|
|  T008|  East|   Kolkata|Store-05| Mobile|2024-01-02| 28000|2024|
|  T009|  East|     Patna|Store-06| Tablet|2024-01-03| 23000|2024|
|  T010|  West|    Mumbai|Store-07| Laptop|2024-01-01| 80000|2024|
|  T011|  West|    Mumbai|Store-07| Mobile|2024-01-02| 35000|2024|
|  T012|  West|      Pune|Store-08| Tablet|2024-01-03| 27000|2

7. Reorder columns in a business-friendly format

In [29]:
df_reordered = df.select("txn_id", "sale_date", "year", "region", "city", "store_id", "product", "amount")
df_reordered.show()

+------+----------+----+------+----------+--------+-------+------+
|txn_id| sale_date|year|region|      city|store_id|product|amount|
+------+----------+----+------+----------+--------+-------+------+
|  T001|2024-01-01|2024| North|     Delhi|Store-01| Laptop| 75000|
|  T002|2024-01-02|2024| North|     Delhi|Store-01| Mobile| 32000|
|  T003|2024-01-03|2024| North|Chandigarh|Store-02| Tablet| 26000|
|  T004|2024-01-01|2024| South| Bangalore|Store-03| Laptop| 78000|
|  T005|2024-01-02|2024| South|   Chennai|Store-04| Mobile| 30000|
|  T006|2024-01-03|2024| South| Bangalore|Store-03| Tablet| 24000|
|  T007|2024-01-01|2024|  East|   Kolkata|Store-05| Laptop| 72000|
|  T008|2024-01-02|2024|  East|   Kolkata|Store-05| Mobile| 28000|
|  T009|2024-01-03|2024|  East|     Patna|Store-06| Tablet| 23000|
|  T010|2024-01-01|2024|  West|    Mumbai|Store-07| Laptop| 80000|
|  T011|2024-01-02|2024|  West|    Mumbai|Store-07| Mobile| 35000|
|  T012|2024-01-03|2024|  West|      Pune|Store-08| Tablet| 27

1. Filter transactions where amount > 50000

In [31]:
df_reordered.filter(df_reordered.amount > 50000).show()

+------+----------+----+------+---------+--------+-------+------+
|txn_id| sale_date|year|region|     city|store_id|product|amount|
+------+----------+----+------+---------+--------+-------+------+
|  T001|2024-01-01|2024| North|    Delhi|Store-01| Laptop| 75000|
|  T004|2024-01-01|2024| South|Bangalore|Store-03| Laptop| 78000|
|  T007|2024-01-01|2024|  East|  Kolkata|Store-05| Laptop| 72000|
|  T010|2024-01-01|2024|  West|   Mumbai|Store-07| Laptop| 80000|
|  T013|2024-01-04|2024| North|    Delhi|Store-01| Laptop| 76000|
|  T014|2024-01-04|2024| South|  Chennai|Store-04| Laptop| 79000|
|  T016|2024-01-04|2024|  West|     Pune|Store-08| Laptop| 77000|
|  T023|2024-01-06|2024|  East|    Patna|Store-06| Laptop| 74000|
+------+----------+----+------+---------+--------+-------+------+



2. Filter only Laptop sales

In [32]:
df_reordered.filter(df_reordered.product == "Laptop").show()

+------+----------+----+------+---------+--------+-------+------+
|txn_id| sale_date|year|region|     city|store_id|product|amount|
+------+----------+----+------+---------+--------+-------+------+
|  T001|2024-01-01|2024| North|    Delhi|Store-01| Laptop| 75000|
|  T004|2024-01-01|2024| South|Bangalore|Store-03| Laptop| 78000|
|  T007|2024-01-01|2024|  East|  Kolkata|Store-05| Laptop| 72000|
|  T010|2024-01-01|2024|  West|   Mumbai|Store-07| Laptop| 80000|
|  T013|2024-01-04|2024| North|    Delhi|Store-01| Laptop| 76000|
|  T014|2024-01-04|2024| South|  Chennai|Store-04| Laptop| 79000|
|  T016|2024-01-04|2024|  West|     Pune|Store-08| Laptop| 77000|
|  T023|2024-01-06|2024|  East|    Patna|Store-06| Laptop| 74000|
+------+----------+----+------+---------+--------+-------+------+



3. Filter sales from North and South regions


In [35]:
df_reordered.filter((df_reordered.region == "North") | (df_reordered.region == "South")).show()

+------+----------+----+------+----------+--------+-------+------+
|txn_id| sale_date|year|region|      city|store_id|product|amount|
+------+----------+----+------+----------+--------+-------+------+
|  T001|2024-01-01|2024| North|     Delhi|Store-01| Laptop| 75000|
|  T002|2024-01-02|2024| North|     Delhi|Store-01| Mobile| 32000|
|  T003|2024-01-03|2024| North|Chandigarh|Store-02| Tablet| 26000|
|  T004|2024-01-01|2024| South| Bangalore|Store-03| Laptop| 78000|
|  T005|2024-01-02|2024| South|   Chennai|Store-04| Mobile| 30000|
|  T006|2024-01-03|2024| South| Bangalore|Store-03| Tablet| 24000|
|  T013|2024-01-04|2024| North|     Delhi|Store-01| Laptop| 76000|
|  T014|2024-01-04|2024| South|   Chennai|Store-04| Laptop| 79000|
|  T017|2024-01-05|2024| North|Chandigarh|Store-02| Mobile| 31000|
|  T018|2024-01-05|2024| South| Bangalore|Store-03| Mobile| 34000|
|  T021|2024-01-06|2024| North|     Delhi|Store-01| Tablet| 28000|
|  T022|2024-01-06|2024| South|   Chennai|Store-04| Tablet| 26

4. Filter sales between 25000 and 75000


In [36]:
df_reordered.filter((df_reordered.amount > 25000) & (df_reordered.amount < 75000)).show()

+------+----------+----+------+----------+--------+-------+------+
|txn_id| sale_date|year|region|      city|store_id|product|amount|
+------+----------+----+------+----------+--------+-------+------+
|  T002|2024-01-02|2024| North|     Delhi|Store-01| Mobile| 32000|
|  T003|2024-01-03|2024| North|Chandigarh|Store-02| Tablet| 26000|
|  T005|2024-01-02|2024| South|   Chennai|Store-04| Mobile| 30000|
|  T007|2024-01-01|2024|  East|   Kolkata|Store-05| Laptop| 72000|
|  T008|2024-01-02|2024|  East|   Kolkata|Store-05| Mobile| 28000|
|  T011|2024-01-02|2024|  West|    Mumbai|Store-07| Mobile| 35000|
|  T012|2024-01-03|2024|  West|      Pune|Store-08| Tablet| 27000|
|  T015|2024-01-04|2024|  East|     Patna|Store-06| Mobile| 29000|
|  T017|2024-01-05|2024| North|Chandigarh|Store-02| Mobile| 31000|
|  T018|2024-01-05|2024| South| Bangalore|Store-03| Mobile| 34000|
|  T020|2024-01-05|2024|  West|    Mumbai|Store-07| Tablet| 29000|
|  T021|2024-01-06|2024| North|     Delhi|Store-01| Tablet| 28

5. Filter transactions from Delhi stores only


In [37]:
df_reordered.filter(df_reordered.city == "Delhi").show()

+------+----------+----+------+-----+--------+-------+------+
|txn_id| sale_date|year|region| city|store_id|product|amount|
+------+----------+----+------+-----+--------+-------+------+
|  T001|2024-01-01|2024| North|Delhi|Store-01| Laptop| 75000|
|  T002|2024-01-02|2024| North|Delhi|Store-01| Mobile| 32000|
|  T013|2024-01-04|2024| North|Delhi|Store-01| Laptop| 76000|
|  T021|2024-01-06|2024| North|Delhi|Store-01| Tablet| 28000|
+------+----------+----+------+-----+--------+-------+------+



6. Apply multiple filters using both filter and where

In [38]:
df_reordered.filter(df_reordered.product == "Laptop").where(df_reordered.region == "North").show()

+------+----------+----+------+-----+--------+-------+------+
|txn_id| sale_date|year|region| city|store_id|product|amount|
+------+----------+----+------+-----+--------+-------+------+
|  T001|2024-01-01|2024| North|Delhi|Store-01| Laptop| 75000|
|  T013|2024-01-04|2024| North|Delhi|Store-01| Laptop| 76000|
+------+----------+----+------+-----+--------+-------+------+



7. Change the order of filters and compare explain(True)



In [39]:
print("Filter Product then Region:")
df_reordered.filter(df_reordered.product == "Laptop").filter(df_reordered.region == "North").explain(True)

print("\nFilter Region then Product:")
df_reordered.filter(df_reordered.region == "North").filter(df_reordered.product == "Laptop").explain(True)

Filter Product then Region:
== Parsed Logical Plan ==
'Filter '`=`(region#40, North)
+- Filter (product#43 = Laptop)
   +- Project [txn_id#39, sale_date#44, year#201, region#40, city#41, store_id#42, product#43, amount#45L]
      +- Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, substring(sale_date#44, 0, 4) AS year#201]
         +- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
txn_id: string, sale_date: string, year: string, region: string, city: string, store_id: string, product: string, amount: bigint
Filter (region#40 = North)
+- Filter (product#43 = Laptop)
   +- Project [txn_id#39, sale_date#44, year#201, region#40, city#41, store_id#42, product#43, amount#45L]
      +- Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, substring(sale_date#44, 0, 4) AS year#201]
         +- LogicalRDD [txn_id#39, region#40, city#41

8. Identify which filters Spark pushes down

In [40]:
# As observed in the 'Optimized Logical Plan' and 'Physical Plan' sections of the previous explain(True) output,
# Spark pushed down both filter conditions together.
# The combined filter shown was: `((isnotnull(product#X) AND isnotnull(region#Y)) AND ((product#X = Laptop) AND (region#Y = North)))`.
# This indicates that Spark's Catalyst Optimizer combined the two separate filter calls into a single, efficient filter operation
# that is applied directly when reading the data from the RDD (Scan ExistingRDD).

1. Total sales amount per region


In [ ]:
from pyspark.sql.functions import sum

df_sales.groupBy("region").agg(sum("amount").alias("total_sales_amount")).show()

### Total sales amount per region using `reduceByKey`

To use `reduceByKey`, we first convert the DataFrame to an RDD and then transform the data into key-value pairs (region, amount) before applying the reduction.

In [46]:
sales_rdd = df_sales.rdd.map(lambda row: (row['region'], row['amount']))

total_sales_by_region_rdd = sales_rdd.reduceByKey(lambda a, b: a + b)

total_sales_by_region_rdd.collect()



[('North', 268000), ('West', 281000), ('South', 271000), ('East', 251000)]

In [43]:
from pyspark.sql.functions import sum

df_sales.groupBy("region").agg(sum("amount").alias("total_sales_amount")).show()

+------+------------------+
|region|total_sales_amount|
+------+------------------+
| South|            271000|
|  East|            251000|
|  West|            281000|
| North|            268000|
+------+------------------+



2. Average sales amount per product


In [47]:
from pyspark.sql.functions import sum

df_sales.groupBy("product").agg(sum("amount").alias("total_sales_per_product")).show()

+-------+-----------------------+
|product|total_sales_per_product|
+-------+-----------------------+
| Laptop|                 611000|
| Mobile|                 252000|
| Tablet|                 208000|
+-------+-----------------------+



3. Maximum sale per city


In [51]:
highest = sales_rdd.reduce(
    lambda x,y: x if x[1] > y[1] else y
)
highest

('Mumbai', 80000)

4. Minimum sale per store


In [52]:
sales_rdd = df_sales.rdd.map(lambda row: (row['city'], row['amount']))
lowest = sales_rdd.reduce(
    lambda x,y: x if x[1] < y[1] else y
)
lowest

('Patna', 23000)

5. Count of transactions per region


In [53]:
df_sales.groupBy("region").count().show()

+------+-----+
|region|count|
+------+-----+
| South|    6|
|  East|    6|
|  West|    6|
| North|    6|
+------+-----+



6. Total revenue per store


In [55]:
df_sales.groupBy("city").sum().show()

+----------+-----------+
|      city|sum(amount)|
+----------+-----------+
| Bangalore|     136000|
|     Patna|     126000|
|   Chennai|     135000|
|    Mumbai|     144000|
|   Kolkata|     125000|
|      Pune|     137000|
|     Delhi|     211000|
|Chandigarh|      57000|
+----------+-----------+



7. Region-wise product sales count



In [56]:
sales_rdd = df_sales.rdd.map(lambda row: (row['product'],row['region'], row['amount']))
df_sales.groupBy("product").count().show()

+-------+-----+
|product|count|
+-------+-----+
| Laptop|    8|
| Mobile|    8|
| Tablet|    8|
+-------+-----+



8. Average transaction value per city

In [59]:
from pyspark.sql.functions import avg, round

df_sales.groupBy("city").agg(round(avg("amount"), 2)).show()

+----------+---------------------+
|      city|round(avg(amount), 2)|
+----------+---------------------+
| Bangalore|             45333.33|
|     Patna|              42000.0|
|   Chennai|              45000.0|
|    Mumbai|              48000.0|
|   Kolkata|             41666.67|
|      Pune|             45666.67|
|     Delhi|              52750.0|
|Chandigarh|              28500.0|
+----------+---------------------+



9. Identify regions with total sales above a threshold[link text]

In [60]:
from pyspark.sql.functions import sum

sales_threshold = 270000

df_sales.groupBy("region").agg(sum("amount").alias("total_sales")) \
    .filter(col("total_sales") > sales_threshold).show()

+------+-----------+
|region|total_sales|
+------+-----------+
| South|     271000|
|  West|     281000|
+------+-----------+



10. Use explain(True) and identify shuffle stages


In [61]:
from pyspark.sql.functions import sum

df_sales.groupBy("region").agg(sum("amount").alias("total_sales_amount")).explain(True)

== Parsed Logical Plan ==
'Aggregate ['region], ['region, 'sum('amount) AS total_sales_amount#635]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
region: string, total_sales_amount: bigint
Aggregate [region#40], [region#40, sum(amount#45L) AS total_sales_amount#635L]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Optimized Logical Plan ==
Aggregate [region#40], [region#40, sum(amount#45L) AS total_sales_amount#635L]
+- Project [region#40, amount#45L]
   +- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- HashAggregate(keys=[region#40], functions=[sum(amount#45L)], output=[region#40, total_sales_amount#635L])
   +- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=852]
      +- HashAggregate(keys=[region#40], 

1. Region + Product wise total sales


In [63]:
df_sales.groupBy("region","product").agg(sum("amount")).show()

+------+-------+-----------+
|region|product|sum(amount)|
+------+-------+-----------+
| North| Laptop|     151000|
| North| Tablet|      54000|
|  East| Tablet|      48000|
|  East| Laptop|     146000|
| South| Tablet|      50000|
| North| Mobile|      63000|
|  West| Tablet|      56000|
|  East| Mobile|      57000|
| South| Mobile|      64000|
| South| Laptop|     157000|
|  West| Mobile|      68000|
|  West| Laptop|     157000|
+------+-------+-----------+



2. City + Store wise average sales


In [64]:
df_sales.groupBy("city","store_id").agg(avg("amount")).show()

+----------+--------+------------------+
|      city|store_id|       avg(amount)|
+----------+--------+------------------+
| Bangalore|Store-03|45333.333333333336|
|     Patna|Store-06|           42000.0|
|   Chennai|Store-04|           45000.0|
|      Pune|Store-08|45666.666666666664|
|Chandigarh|Store-02|           28500.0|
|   Kolkata|Store-05|41666.666666666664|
|    Mumbai|Store-07|           48000.0|
|     Delhi|Store-01|           52750.0|
+----------+--------+------------------+



3. Region + City wise transaction count


In [65]:
df_sales.groupBy("region","city").count().show()

+------+----------+-----+
|region|      city|count|
+------+----------+-----+
|  West|    Mumbai|    3|
| South| Bangalore|    3|
| North|     Delhi|    4|
| North|Chandigarh|    2|
| South|   Chennai|    3|
|  West|      Pune|    3|
|  East|   Kolkata|    3|
|  East|     Patna|    3|
+------+----------+-----+



4. Product + Store wise max sale


In [66]:
df_sales.groupBy("region","city").max().show()

+------+----------+-----------+
|region|      city|max(amount)|
+------+----------+-----------+
|  West|    Mumbai|      80000|
| South| Bangalore|      78000|
| North|     Delhi|      76000|
| North|Chandigarh|      31000|
| South|   Chennai|      79000|
|  West|      Pune|      77000|
|  East|   Kolkata|      72000|
|  East|     Patna|      74000|
+------+----------+-----------+



5. Identify top-selling product per region using aggregation only

In [81]:
from pyspark.sql import functions as f
df_c=df_sales.groupBy("region","product").agg(f.sum("amount").alias("max_total"))
df_c.show()

df_top = (
    df_c.groupBy("region").agg(f.max(f.struct("max_total", "product"))
))
df_top.show()


+------+-------+---------+
|region|product|max_total|
+------+-------+---------+
| North| Laptop|   151000|
| North| Tablet|    54000|
|  East| Tablet|    48000|
|  East| Laptop|   146000|
| South| Tablet|    50000|
| North| Mobile|    63000|
|  West| Tablet|    56000|
|  East| Mobile|    57000|
| South| Mobile|    64000|
| South| Laptop|   157000|
|  West| Mobile|    68000|
|  West| Laptop|   157000|
+------+-------+---------+

+------+-------------------------------+
|region|max(struct(max_total, product))|
+------+-------------------------------+
|  East|               {146000, Laptop}|
| North|               {151000, Laptop}|
| South|               {157000, Laptop}|
|  West|               {157000, Laptop}|
+------+-------------------------------+



In [ ]:
from pyspark.sql.window import Window
from pyspark.sql.functions import sum, rank, row_number, dense_rank

In [ ]:
from pyspark.sql.window import Window
from pyspark.sql.functions import sum, rank, row_number, dense_rank

In [68]:
from pyspark.sql.window import Window
from pyspark.sql.functions import sum, rank, row_number, dense_rank

1. Compute running total of sales per region ordered by date


In [73]:
from pyspark.sql.window import Window
from pyspark.sql.functions import sum, col

window_spec = Window.partitionBy("region").orderBy("sale_date")
df_sales.withColumn("running_total_sales", sum("amount").over(window_spec)).show()

+------+------+----------+--------+-------+----------+------+-------------------+
|txn_id|region|      city|store_id|product| sale_date|amount|running_total_sales|
+------+------+----------+--------+-------+----------+------+-------------------+
|  T007|  East|   Kolkata|Store-05| Laptop|2024-01-01| 72000|              72000|
|  T008|  East|   Kolkata|Store-05| Mobile|2024-01-02| 28000|             100000|
|  T009|  East|     Patna|Store-06| Tablet|2024-01-03| 23000|             123000|
|  T015|  East|     Patna|Store-06| Mobile|2024-01-04| 29000|             152000|
|  T019|  East|   Kolkata|Store-05| Tablet|2024-01-05| 25000|             177000|
|  T023|  East|     Patna|Store-06| Laptop|2024-01-06| 74000|             251000|
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|              75000|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|             107000|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|             133000|
|  T013| North| 

2. Rank transactions by amount within each region


In [82]:
from pyspark.sql.functions import rank
window_spec = Window.partitionBy("region").orderBy("amount")
ranked_df = df.withColumn("Rank", rank().over(window_spec))
ranked_df.show()

+------+------+----------+--------+-------+----------+------+----+----+
|txn_id|region|      city|store_id|product| sale_date|amount|year|Rank|
+------+------+----------+--------+-------+----------+------+----+----+
|  T009|  East|     Patna|Store-06| Tablet|2024-01-03| 23000|2024|   1|
|  T019|  East|   Kolkata|Store-05| Tablet|2024-01-05| 25000|2024|   2|
|  T008|  East|   Kolkata|Store-05| Mobile|2024-01-02| 28000|2024|   3|
|  T015|  East|     Patna|Store-06| Mobile|2024-01-04| 29000|2024|   4|
|  T007|  East|   Kolkata|Store-05| Laptop|2024-01-01| 72000|2024|   5|
|  T023|  East|     Patna|Store-06| Laptop|2024-01-06| 74000|2024|   6|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|2024|   1|
|  T021| North|     Delhi|Store-01| Tablet|2024-01-06| 28000|2024|   2|
|  T017| North|Chandigarh|Store-02| Mobile|2024-01-05| 31000|2024|   3|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|2024|   4|
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|202

3. Assign row numbers per store ordered by sale amount


In [83]:
window_spec = Window.partitionBy("store_id").orderBy("amount")
ranked_df = df.withColumn("Rank", rank().over(window_spec))
ranked_df.show()

+------+------+----------+--------+-------+----------+------+----+----+
|txn_id|region|      city|store_id|product| sale_date|amount|year|Rank|
+------+------+----------+--------+-------+----------+------+----+----+
|  T021| North|     Delhi|Store-01| Tablet|2024-01-06| 28000|2024|   1|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|2024|   2|
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|2024|   3|
|  T013| North|     Delhi|Store-01| Laptop|2024-01-04| 76000|2024|   4|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|2024|   1|
|  T017| North|Chandigarh|Store-02| Mobile|2024-01-05| 31000|2024|   2|
|  T006| South| Bangalore|Store-03| Tablet|2024-01-03| 24000|2024|   1|
|  T018| South| Bangalore|Store-03| Mobile|2024-01-05| 34000|2024|   2|
|  T004| South| Bangalore|Store-03| Laptop|2024-01-01| 78000|2024|   3|
|  T022| South|   Chennai|Store-04| Tablet|2024-01-06| 26000|2024|   1|
|  T005| South|   Chennai|Store-04| Mobile|2024-01-02| 30000|202

4. Use dense rank to rank products per region


In [85]:
from pyspark.sql.functions import dense_rank
window_spec = Window.partitionBy("region").orderBy(col("product").desc())

# Apply dense_rank
ranked_df = df.withColumn("dense_rank", dense_rank().over(window_spec))
ranked_df.show()

+------+------+----------+--------+-------+----------+------+----+----------+
|txn_id|region|      city|store_id|product| sale_date|amount|year|dense_rank|
+------+------+----------+--------+-------+----------+------+----+----------+
|  T009|  East|     Patna|Store-06| Tablet|2024-01-03| 23000|2024|         1|
|  T019|  East|   Kolkata|Store-05| Tablet|2024-01-05| 25000|2024|         1|
|  T008|  East|   Kolkata|Store-05| Mobile|2024-01-02| 28000|2024|         2|
|  T015|  East|     Patna|Store-06| Mobile|2024-01-04| 29000|2024|         2|
|  T007|  East|   Kolkata|Store-05| Laptop|2024-01-01| 72000|2024|         3|
|  T023|  East|     Patna|Store-06| Laptop|2024-01-06| 74000|2024|         3|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|2024|         1|
|  T021| North|     Delhi|Store-01| Tablet|2024-01-06| 28000|2024|         1|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|2024|         2|
|  T017| North|Chandigarh|Store-02| Mobile|2024-01-05| 31000|202

5. Identify top 2 highest sales per region using window functions


In [88]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number, col

window_spec = Window.partitionBy("region").orderBy(col("amount").desc())
ranked_df = df.withColumn("row_num", row_number().over(window_spec))
ranked_df.filter(col("row_num") <= 2).show()

+------+------+---------+--------+-------+----------+------+----+-------+
|txn_id|region|     city|store_id|product| sale_date|amount|year|row_num|
+------+------+---------+--------+-------+----------+------+----+-------+
|  T023|  East|    Patna|Store-06| Laptop|2024-01-06| 74000|2024|      1|
|  T007|  East|  Kolkata|Store-05| Laptop|2024-01-01| 72000|2024|      2|
|  T013| North|    Delhi|Store-01| Laptop|2024-01-04| 76000|2024|      1|
|  T001| North|    Delhi|Store-01| Laptop|2024-01-01| 75000|2024|      2|
|  T014| South|  Chennai|Store-04| Laptop|2024-01-04| 79000|2024|      1|
|  T004| South|Bangalore|Store-03| Laptop|2024-01-01| 78000|2024|      2|
|  T010|  West|   Mumbai|Store-07| Laptop|2024-01-01| 80000|2024|      1|
|  T016|  West|     Pune|Store-08| Laptop|2024-01-04| 77000|2024|      2|
+------+------+---------+--------+-------+----------+------+----+-------+



6. Compare rank vs dense_rank output


In [86]:
window_spec_amount = Window.partitionBy("region").orderBy("amount")

# Calculate rank
df_with_rank = df.withColumn("rank_output", rank().over(window_spec_amount))

# Calculate dense_rank
df_with_dense_rank = df_with_rank.withColumn("dense_rank_output", dense_rank().over(window_spec_amount))

# Show the comparison for relevant columns
df_with_dense_rank.select("region", "product", "amount", "rank_output", "dense_rank_output").orderBy("region", "amount").show()

+------+-------+------+-----------+-----------------+
|region|product|amount|rank_output|dense_rank_output|
+------+-------+------+-----------+-----------------+
|  East| Tablet| 23000|          1|                1|
|  East| Tablet| 25000|          2|                2|
|  East| Mobile| 28000|          3|                3|
|  East| Mobile| 29000|          4|                4|
|  East| Laptop| 72000|          5|                5|
|  East| Laptop| 74000|          6|                6|
| North| Tablet| 26000|          1|                1|
| North| Tablet| 28000|          2|                2|
| North| Mobile| 31000|          3|                3|
| North| Mobile| 32000|          4|                4|
| North| Laptop| 75000|          5|                5|
| North| Laptop| 76000|          6|                6|
| South| Tablet| 24000|          1|                1|
| South| Tablet| 26000|          2|                2|
| South| Mobile| 30000|          3|                3|
| South| Mobile| 34000|     

7. Calculate cumulative sales per store


In [90]:
window_spec = Window.partitionBy("store_id")
ranked_df = df.withColumn("sum", sum("amount").over(window_spec))
ranked_df.show()

+------+------+----------+--------+-------+----------+------+----+------+
|txn_id|region|      city|store_id|product| sale_date|amount|year|   sum|
+------+------+----------+--------+-------+----------+------+----+------+
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|2024|211000|
|  T002| North|     Delhi|Store-01| Mobile|2024-01-02| 32000|2024|211000|
|  T013| North|     Delhi|Store-01| Laptop|2024-01-04| 76000|2024|211000|
|  T021| North|     Delhi|Store-01| Tablet|2024-01-06| 28000|2024|211000|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|2024| 57000|
|  T017| North|Chandigarh|Store-02| Mobile|2024-01-05| 31000|2024| 57000|
|  T004| South| Bangalore|Store-03| Laptop|2024-01-01| 78000|2024|136000|
|  T006| South| Bangalore|Store-03| Tablet|2024-01-03| 24000|2024|136000|
|  T018| South| Bangalore|Store-03| Mobile|2024-01-05| 34000|2024|136000|
|  T005| South|   Chennai|Store-04| Mobile|2024-01-02| 30000|2024|135000|
|  T014| South|   Chennai|Store-04| La

8. Identify first and last transaction per city using windows

In [91]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number, col

# Window specification for the first transaction (ordered by sale_date ascending)
window_first = Window.partitionBy("city").orderBy("sale_date")

# Add row number for first transaction and filter
df_with_first_rn = df_sales.withColumn("row_num_asc", row_number().over(window_first))
first_transactions = df_with_first_rn.filter(col("row_num_asc") == 1).drop("row_num_asc")

print("First transactions per city:")
first_transactions.show()

# Window specification for the last transaction (ordered by sale_date descending)
window_last = Window.partitionBy("city").orderBy(col("sale_date").desc())

# Add row number for last transaction and filter
df_with_last_rn = df_sales.withColumn("row_num_desc", row_number().over(window_last))
last_transactions = df_with_last_rn.filter(col("row_num_desc") == 1).drop("row_num_desc")

print("Last transactions per city:")
last_transactions.show()

First transactions per city:
+------+------+----------+--------+-------+----------+------+
|txn_id|region|      city|store_id|product| sale_date|amount|
+------+------+----------+--------+-------+----------+------+
|  T004| South| Bangalore|Store-03| Laptop|2024-01-01| 78000|
|  T003| North|Chandigarh|Store-02| Tablet|2024-01-03| 26000|
|  T005| South|   Chennai|Store-04| Mobile|2024-01-02| 30000|
|  T001| North|     Delhi|Store-01| Laptop|2024-01-01| 75000|
|  T007|  East|   Kolkata|Store-05| Laptop|2024-01-01| 72000|
|  T010|  West|    Mumbai|Store-07| Laptop|2024-01-01| 80000|
|  T009|  East|     Patna|Store-06| Tablet|2024-01-03| 23000|
|  T012|  West|      Pune|Store-08| Tablet|2024-01-03| 27000|
+------+------+----------+--------+-------+----------+------+

Last transactions per city:
+------+------+----------+--------+-------+----------+------+
|txn_id|region|      city|store_id|product| sale_date|amount|
+------+------+----------+--------+-------+----------+------+
|  T018| Sou

# Task
The original user task is to:
"Analyze and explain the physical execution plans of various Apache Spark DataFrame operations, specifically `select`, `filter`, `groupBy`, and `window functions`, identifying the presence of shuffles, exchanges, and sorts, and providing an explanation for why window functions introduce sorting."

## Explain Select Operation

### Subtask:
Run explain(True) for a simple DataFrame select operation to observe its physical plan.


**Reasoning**:
To observe the physical plan of a simple DataFrame select operation, I will select the specified columns from the `df_sales` DataFrame and then call the `explain(True)` method on the result.



In [92]:
df_sales.select("region", "product", "amount").explain(True)

== Parsed Logical Plan ==
'Project ['region, 'product, 'amount]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
region: string, product: string, amount: bigint
Project [region#40, product#43, amount#45L]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Optimized Logical Plan ==
Project [region#40, product#43, amount#45L]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Physical Plan ==
*(1) Project [region#40, product#43, amount#45L]
+- *(1) Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]



### Explanation of Select Operation's Physical Plan

The `explain(True)` output for the simple `select` operation (`df_sales.select("region", "product", "amount")`) shows a straightforward execution plan without any complex operations like shuffles, exchanges, or sorts.

**Physical Plan Breakdown:**

1.  **`*(1) Project [region#40, product#43, amount#45L]`**: This indicates that Spark is performing a projection operation. It's essentially selecting (or picking out) the specified columns (`region`, `product`, `amount`) from the input data. This is a narrow transformation, meaning it doesn't require shuffling data across partitions or nodes.

2.  **`+- *(1) Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: This is the source of the data. Spark is scanning an existing Resilient Distributed Dataset (RDD), which in this case was created from the in-memory `sales_data`. This step involves reading the raw data.

**Absence of Shuffles, Exchanges, and Sorts:**

*   **Shuffles/Exchanges**: These operations are typically introduced by wide transformations such as `groupByKey`, `reduceByKey`, `join`, or `repartition`. A `select` operation only involves picking columns and doesn't change the partitioning of the data or require data movement between executors. Therefore, no shuffles or exchanges are necessary.
*   **Sorts**: Sorting operations are introduced by `orderBy` clauses or certain window functions. Since there is no explicit sorting requested in a simple `select` statement, no sort operations appear in the physical plan.

In summary, the physical plan for a `select` operation is very efficient as it only involves reading the data and projecting the required columns, all within the existing partitions, hence avoiding costly data movement operations.

## Explain Filter Operation

### Subtask:
Run `explain(True)` for a DataFrame filter operation to observe its physical plan.

#### Instructions
1. Filter the `df_reordered` DataFrame to include only sales where the product is "Laptop".
2. Call the `.explain(True)` method on the resulting DataFrame to display its physical execution plan.

**Reasoning**:
As per the instructions, I will now filter the `df_reordered` DataFrame for 'Laptop' products and then apply `explain(True)` to display its physical plan.



In [93]:
df_reordered.filter(df_reordered.product == "Laptop").explain(True)

== Parsed Logical Plan ==
'Filter '`=`(product#43, Laptop)
+- Project [txn_id#39, sale_date#44, year#201, region#40, city#41, store_id#42, product#43, amount#45L]
   +- Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, substring(sale_date#44, 0, 4) AS year#201]
      +- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
txn_id: string, sale_date: string, year: string, region: string, city: string, store_id: string, product: string, amount: bigint
Filter (product#43 = Laptop)
+- Project [txn_id#39, sale_date#44, year#201, region#40, city#41, store_id#42, product#43, amount#45L]
   +- Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, substring(sale_date#44, 0, 4) AS year#201]
      +- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Optimized Logical Plan ==
Project [txn_id#

### Explanation of Filter Operation's Physical Plan

The `explain(True)` output for the `filter` operation (`df_reordered.filter(df_reordered.product == "Laptop")`) reveals that Spark processes filters efficiently, similar to `select`, without introducing costly operations like shuffles, exchanges, or explicit sorts.

**Physical Plan Breakdown:**

1.  **`*(1) Project [txn_id#39, sale_date#44, substring(sale_date#44, 0, 4) AS year#201, region#40, city#41, store_id#42, product#43, amount#45L]`**: This is a projection step that selects and potentially transforms columns. In this specific case, the `year` column was derived earlier, and this step ensures all relevant columns are retained in the final output.

2.  **`+- *(1) Filter (isnotnull(product#43) AND (product#43 = Laptop))`**: This is the core filtering operation. Spark applies the condition `product == "Laptop"` to each row. The `isnotnull(product#43)` part is added by Spark's optimizer to handle potential null values gracefully, ensuring that only non-null 'product' values are compared. This operation is a **narrow transformation** because each row can be processed independently to determine if it meets the filter criteria. The data within partitions does not need to be rearranged or moved to other partitions.

3.  **`+- *(1) Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: This is the initial step where Spark reads the data from the underlying RDD (which represents our `df_sales` DataFrame). The filter condition is pushed down as close to the data source as possible (predicate pushdown), allowing Spark to read only the necessary data from the outset.

**Absence of Shuffles, Exchanges, and Sorts:**

*   **Shuffles/Exchanges**: Filter operations, being narrow transformations, do not require data to be redistributed across the network. Each partition can be filtered independently on the node where it resides. Thus, there are no `Exchange` operations (which signify shuffles).
*   **Sorts**: No explicit sorting was requested in the filter operation, and filtering itself does not inherently require data to be sorted. Therefore, no `Sort` operations appear in the physical plan.

Spark's Catalyst Optimizer intelligently combines the projection and filter operations and pushes the filter predicate down to the data source, optimizing performance by reducing the amount of data processed and transferred.

## Explain GroupBy Operation

### Subtask:
Run `explain(True)` for a DataFrame `groupBy` and `agg` operation to observe its physical plan.

#### Instructions
1. Group the `df_sales` DataFrame by `region`.
2. Aggregate the sum of `amount` and alias it as `total_sales_amount`.
3. Call the `.explain(True)` method on the resulting DataFrame to display its physical execution plan.

**Reasoning**:
As per the instructions in the preceding markdown block, I will now group the `df_sales` DataFrame by `region`, aggregate the sum of `amount`, and then apply `explain(True)` to display its physical plan.



In [94]:
from pyspark.sql.functions import sum

df_sales.groupBy("region").agg(sum("amount").alias("total_sales_amount")).explain(True)

== Parsed Logical Plan ==
'Aggregate ['region], ['region, 'sum('amount) AS total_sales_amount#1314]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
region: string, total_sales_amount: bigint
Aggregate [region#40], [region#40, sum(amount#45L) AS total_sales_amount#1314L]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Optimized Logical Plan ==
Aggregate [region#40], [region#40, sum(amount#45L) AS total_sales_amount#1314L]
+- Project [region#40, amount#45L]
   +- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- HashAggregate(keys=[region#40], functions=[sum(amount#45L)], output=[region#40, total_sales_amount#1314L])
   +- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=2474]
      +- HashAggregate(keys=[region#

### Explanation of GroupBy Operation's Physical Plan

The `explain(True)` output for the `groupBy` and `agg` operation (`df_sales.groupBy("region").agg(sum("amount").alias("total_sales_amount"))`) shows a more complex physical plan compared to `select` and `filter`, primarily due to the introduction of **shuffles** (represented by `Exchange`) and **aggregate** stages.

**Physical Plan Breakdown:**

1.  **`AdaptiveSparkPlan isFinalPlan=false`**: This indicates that Spark's Adaptive Query Execution (AQE) is enabled, which can dynamically optimize the query plan during execution based on runtime statistics.

2.  **`+- HashAggregate(keys=[region#40], functions=[sum(amount#45L)], output=[region#40, total_sales_amount#1314L])`**: This is the final aggregation step. After all the data for each `region` has been brought together, Spark performs the final `sum` aggregation to compute the `total_sales_amount` for each unique region. This is typically the second stage of a two-stage aggregation.

3.  **`+- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=2474]`**: This is the crucial step that introduces a **shuffle**. For a `groupBy` operation, all records belonging to the same grouping key (in this case, `region`) must be processed together to compute the aggregate. If data for a given region is spread across different partitions on different executors, Spark needs to redistribute (shuffle) the data so that all data for a specific `region` ends up on the same partition. `hashpartitioning(region#40, 200)` means Spark is using a hash function on the `region` column to determine which of the 200 default partitions each row should go to. `ENSURE_REQUIREMENTS` ensures that the data meets the partitioning requirements for the subsequent aggregation step.

4.  **`+- HashAggregate(keys=[region#40], functions=[partial_sum(amount#45L)], output=[region#40, sum#1324L])`**: This is the **partial aggregation** step, performed *before* the shuffle. To optimize performance and reduce the amount of data shuffled across the network, Spark often performs a local (partial) aggregation on each partition first. For example, if a partition has multiple sales for 'North' region, Spark will sum these locally before sending them over the network. This reduces the data volume that needs to be shuffled.

5.  **`+- Project [region#40, amount#45L]`**: This projects only the necessary columns (`region` and `amount`) from the input for the aggregation, discarding other columns early on.

6.  **`+- Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: The initial step of reading the data from the source RDD.

**Presence of Shuffles (Exchanges):**

The `groupBy` operation inherently requires a **shuffle** (represented by the `Exchange` operator) because it's a **wide transformation**. To compute aggregates like `sum` for each region, all rows corresponding to a particular region must be collected and processed together. If these rows are distributed across different partitions or nodes, Spark must redistribute the data across the cluster to ensure that all values for a given key are on the same machine/partition. This data movement across the network is expensive and is a primary factor impacting the performance of `groupBy` operations.

**Absence of Sorts (in this specific case):**

While some aggregations might implicitly involve sorting (e.g., if you were to use `orderBy` or certain window functions), a simple `groupBy` with `sum` aggregation does not inherently require sorting. Spark uses `HashAggregate` here, which typically involves hashing keys to group them, not sorting. Therefore, no explicit `Sort` operations are present in this physical plan.

## Explain Window Functions Operation

### Subtask:
Run `explain(True)` for a DataFrame operation involving a window function to observe its physical plan, identifying shuffles, exchanges, and sorts, and explaining why sorts are introduced.

#### Instructions
1.  Define a window specification that partitions by `region` and orders by `sale_date`.
2.  Calculate the `running_total_sales` using `sum("amount")` over this window on the `df_sales` DataFrame.
3.  Call the `.explain(True)` method on the resulting DataFrame to display its physical execution plan.

**Reasoning**:
As per the instructions in the preceding markdown block, I will define a window specification, apply it to calculate running total sales, and then call `explain(True)` to display its physical plan.



In [95]:
from pyspark.sql.window import Window
from pyspark.sql.functions import sum, col

window_spec = Window.partitionBy("region").orderBy("sale_date")
df_sales.withColumn("running_total_sales", sum("amount").over(window_spec)).explain(True)

== Parsed Logical Plan ==
'Project [unresolvedstarwithcolumns(running_total_sales, 'sum('amount) windowspecdefinition('region, 'sale_date ASC NULLS FIRST, unspecifiedframe$()), None)]
+- LogicalRDD [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L], false

== Analyzed Logical Plan ==
txn_id: string, region: string, city: string, store_id: string, product: string, sale_date: string, amount: bigint, running_total_sales: bigint
Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, running_total_sales#1325L]
+- Project [txn_id#39, region#40, city#41, store_id#42, product#43, sale_date#44, amount#45L, running_total_sales#1325L, running_total_sales#1325L]
   +- Window [sum(amount#45L) windowspecdefinition(region#40, sale_date#44 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS running_total_sales#1325L], [region#40], [sale_date#44 ASC NULLS FIRST]
      +- Project [txn_id#39, regio

### Explanation of Window Function's Physical Plan

The `explain(True)` output for the window function operation (`df_sales.withColumn("running_total_sales", sum("amount").over(window_spec))`) reveals a physical plan that includes **shuffles**, **sorts**, and the **Window** operator itself. This complexity arises because window functions require data to be grouped (partitioned) and ordered before calculations can be performed within each window.

**Physical Plan Breakdown:**

1.  **`AdaptiveSparkPlan isFinalPlan=false`**: Indicates that Adaptive Query Execution (AQE) is enabled for potential runtime optimizations.

2.  **`+- Window [sum(amount#45L) windowspecdefinition(region#40, sale_date#44 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS running_total_sales#1325L], [region#40], [sale_date#44 ASC NULLS FIRST]`**: This is the actual window operation. It computes the `sum(amount)` for the `running_total_sales` within each defined window. The window is partitioned by `region` and ordered by `sale_date` (ascending, nulls first), with a frame of `unboundedpreceding` to `currentrow` (which defines the running total).

3.  **`+- Sort [region#40 ASC NULLS FIRST, sale_date#44 ASC NULLS FIRST], false, 0`**: This step explicitly introduces a **sort** operation. Data is sorted first by `region` and then by `sale_date` within each `region`. The `false` indicates that this is not a global sort but rather a local sort within partitions after shuffling.

4.  **`+- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=2484]`**: This is a **shuffle** operation. For window functions partitioned by a column (here, `region`), Spark needs to ensure that all rows belonging to the same `region` are grouped together on the same partition. `hashpartitioning(region#40, 200)` redistributes the data based on the hash of the `region` column into 200 partitions. This is a wide transformation, incurring network I/O.

5.  **`+- Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: The initial step of reading the data from the source RDD.

**Why Window Functions Introduce Sorting:**

Window functions operate on a defined

### Explanation of Window Function's Physical Plan

The `explain(True)` output for the window function operation (`df_sales.withColumn("running_total_sales", sum("amount").over(window_spec))`) reveals a physical plan that includes **shuffles**, **sorts**, and the **Window** operator itself. This complexity arises because window functions require data to be grouped (partitioned) and ordered before calculations can be performed within each window.

**Physical Plan Breakdown:**

1.  **`AdaptiveSparkPlan isFinalPlan=false`**: Indicates that Adaptive Query Execution (AQE) is enabled for potential runtime optimizations.

2.  **`+- Window [sum(amount#45L) windowspecdefinition(region#40, sale_date#44 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS running_total_sales#1325L], [region#40], [sale_date#44 ASC NULLS FIRST]`**: This is the actual window operation. It computes the `sum(amount)` for the `running_total_sales` within each defined window. The window is partitioned by `region` and ordered by `sale_date` (ascending, nulls first), with a frame of `unboundedpreceding` to `currentrow` (which defines the running total).

3.  **`+- Sort [region#40 ASC NULLS FIRST, sale_date#44 ASC NULLS FIRST], false, 0`**: This step explicitly introduces a **sort** operation. Data is sorted first by `region` and then by `sale_date` within each `region`. The `false` indicates that this is not a global sort but rather a local sort within partitions after shuffling.

4.  **`+- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=2484]`**: This is a **shuffle** operation. For window functions partitioned by a column (here, `region`), Spark needs to ensure that all rows belonging to the same `region` are grouped together on the same partition. `hashpartitioning(region#40, 200)` redistributes the data based on the hash of the `region` column into 200 partitions. This is a wide transformation, incurring network I/O.

5.  **`+- Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: The initial step of reading the data from the source RDD.

**Why Window Functions Introduce Sorting:**

Window functions operate on a defined **frame of rows** relative to the current row, within each partition. The `ORDER BY` clause within a window specification is fundamental because it dictates the sequence in which these rows are considered. For calculations like a running total (`sum(...) OVER (ORDER BY ...)`), or for ranking functions (`rank()`, `row_number()`), the order of data is critical to produce correct and deterministic results. Spark must physically sort the data within each partition according to the `ORDER BY` clause of the window specification to guarantee that the window frame is correctly established and the aggregate or ranking function operates on the intended set of ordered rows. This ensures that the results (e.g., the sum up to a certain date, or the rank of a transaction) are accurate based on the specified order.

### Explanation of Window Function's Physical Plan

The `explain(True)` output for the window function operation (`df_sales.withColumn("running_total_sales", sum("amount").over(window_spec))`) reveals a physical plan that includes **shuffles**, **sorts**, and the **Window** operator itself. This complexity arises because window functions require data to be grouped (partitioned) and ordered before calculations can be performed within each window.

**Physical Plan Breakdown:**

1.  **`AdaptiveSparkPlan isFinalPlan=false`**: Indicates that Adaptive Query Execution (AQE) is enabled for potential runtime optimizations.

2.  **`+- Window [sum(amount#45L) windowspecdefinition(region#40, sale_date#44 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) AS running_total_sales#1325L], [region#40], [sale_date#44 ASC NULLS FIRST]`**: This is the actual window operation. It computes the `sum(amount)` for the `running_total_sales` within each defined window. The window is partitioned by `region` and ordered by `sale_date` (ascending, nulls first), with a frame of `unboundedpreceding` to `currentrow` (which defines the running total).

3.  **`+- Sort [region#40 ASC NULLS FIRST, sale_date#44 ASC NULLS FIRST], false, 0`**: This step explicitly introduces a **sort** operation. Data is sorted first by `region` and then by `sale_date` within each `region`. The `false` indicates that this is not a global sort but rather a local sort within partitions after shuffling.

4.  **`+- Exchange hashpartitioning(region#40, 200), ENSURE_REQUIREMENTS, [plan_id=2484]`**: This is a **shuffle** operation. For window functions partitioned by a column (here, `region`), Spark needs to ensure that all rows belonging to the same `region` are grouped together on the same partition. `hashpartitioning(region#40, 200)` redistributes the data based on the hash of the `region` column into 200 partitions. This is a wide transformation, incurring network I/O.

5.  **`+- Scan ExistingRDD[txn_id#39,region#40,city#41,store_id#42,product#43,sale_date#44,amount#45L]`**: The initial step of reading the data from the source RDD.

**Why Window Functions Introduce Sorting:**

Window functions operate on a defined **frame of rows** relative to the current row, within each partition. The `ORDER BY` clause within a window specification is fundamental because it dictates the sequence in which these rows are considered. For calculations like a running total (`sum(...) OVER (ORDER BY ...)`), or for ranking functions (`rank()`, `row_number()`), the order of data is critical to produce correct and deterministic results. Spark must physically sort the data within each partition according to the `ORDER BY` clause of the window specification to guarantee that the window frame is correctly established and the aggregate or ranking function operates on the intended set of ordered rows. This ensures that the results (e.g., the sum up to a certain date, or the rank of a transaction) are accurate based on the specified order.

## Analyze Explain Outputs and Explain Window Sorting

### Subtask:
Consolidate the observations from the explain outputs for select, filter, groupBy, and window functions, identify the presence of shuffles, exchanges, and sorts in each, and provide a comprehensive explanation for why window functions introduce sorting.


## Summary:

### Q&A
1.  **Which DataFrame operations introduce shuffles or exchanges?**
    The `groupBy` operation introduces a shuffle (observed as `Exchange hashpartitioning`) because it's a wide transformation requiring data redistribution to group records by key. Similarly, `window functions` also introduce a shuffle (`Exchange hashpartitioning`) when a `partitionBy` clause is specified, to co-locate all rows belonging to the same partition.
2.  **Which DataFrame operations introduce sorts?**
    The `window functions` operation explicitly introduces a sort (observed as `Sort [region#40 ASC NULLS FIRST, sale_date#44 ASC NULLS FIRST]`) due to the `orderBy` clause within the window specification. Simple `select`, `filter`, and `groupBy` (with `HashAggregate`) operations do not introduce sorts.
3.  **Why do window functions introduce sorting?**
    Window functions introduce sorting to ensure deterministic and correct calculations within a defined window frame. The `ORDER BY` clause within a window specification dictates the sequence in which rows are processed, which is critical for functions like running totals or ranking. Spark must physically sort the data within each partition according to this clause to establish the correct window frame and perform accurate computations.

### Data Analysis Key Findings
*   **`select` and `filter` operations are narrow transformations**: They execute efficiently, involving only `Project` and `Filter` operations, respectively, followed by `Scan ExistingRDD`. Neither introduces shuffles, exchanges, or sorts, making them generally inexpensive operations.
*   **`groupBy` operations are wide transformations and introduce shuffles**: The physical plan for `groupBy` includes a `HashAggregate` (partial), followed by an `Exchange hashpartitioning` (shuffle), and then a final `HashAggregate`. The shuffle is necessary to bring all data for a specific grouping key together for aggregation.
*   **`window functions` are complex wide transformations that introduce both shuffles and sorts**: The physical plan for window functions includes an `Exchange hashpartitioning` (shuffle) for partitioning the data by the window's `partitionBy` clause, followed by an explicit `Sort` operation based on the window's `orderBy` clause. Finally, the `Window` operator performs the calculation.
*   The `Sort` operation observed in the window function's plan is directly attributed to the `orderBy("sale_date")` clause, indicating Spark's necessity to order data for accurate window-based calculations.

### Insights or Next Steps
*   **Cost Implications**: Operations introducing `Exchange` (shuffles) and `Sort` are generally more resource-intensive due to network I/O and CPU overhead. When designing Spark applications, minimize such wide transformations where possible.
*   **Optimize Window Functions**: For window functions, consider if the `ORDER BY` clause is strictly necessary or if an alternative approach can achieve the desired result with fewer costly operations, especially on very large datasets.
